src/nvim/os/shell.c

```diff
@@ -281,25 +281,22 @@ static void dynamic_buffer_ensure(DynamicBuffer *buf, size_t desired)
   buf->data = xrealloc(buf->data, buf->cap);
 }
 
-static void system_data_cb(RStream *rstream, void *data, bool eof)
+static void system_data_cb(RStream *rstream, RBuffer *buf, void *data, bool eof)
 {
   Job *job = data;
-  DynamicBuffer *buf = job_data(job);
+  DynamicBuffer *dbuf = job_data(job);
 
-  size_t nread = rstream_pending(rstream);
-
-  dynamic_buffer_ensure(buf, buf->len + nread + 1);
-  rstream_read(rstream, buf->data + buf->len, nread);
-
-  buf->len += nread;
+  size_t nread = buf->size;
+  dynamic_buffer_ensure(dbuf, dbuf->len + nread + 1);
+  rbuffer_read(buf, dbuf->data + dbuf->len, nread);
+  dbuf->len += nread;
 }
 
-static void out_data_cb(RStream *rstream, void *data, bool eof)
+static void out_data_cb(RStream *rstream, RBuffer *buf, void *data, bool eof)
 {
-  RBuffer *rbuffer = rstream_buffer(rstream);
-  size_t written = write_output(rbuffer_read_ptr(rbuffer),
-                                rbuffer_pending(rbuffer), false, eof);
-  rbuffer_consumed(rbuffer, written);
+  RBUFFER_UNTIL_READ_OR_EMPTY(buf, ptr, len) {
+    rbuffer_consumed(buf, write_output(ptr, len, false, eof));
```

Comment:
Does not work for the wrap around case:
If in the first half no `NL` and `NUL` are found, `write_output()` returns `0`. So in the second (and last) iteration of the loop the first half is used again, but the second half is not used (possible `NL` here).
If a `NUL` was found in the first iteration, it would be changed to a `NL` and the next iteration would
output "something".

Is `write_output()` working correctly for `eof == true`, if the two buffer halfs are written not as one?